include/swift/RemoteInspection/TypeRefBuilder.h

```diff
@@ -486,6 +455,368 @@ class TypeRefBuilder {
+
+    template <typename Record, typename Field>
+    RemoteRef<char> readTypeRef(RemoteRef<Record> record, const Field &field) {
+      uint64_t remoteAddr = record.resolveRelativeFieldData(field);
+
+      return readTypeRef(remoteAddr);
+    }
+
+    llvm::Optional<std::string> normalizeReflectionName(RemoteRef<char> name);
+
+  private:
+    /// Get the primitive type lowering for a builtin type.
+    RemoteRef<BuiltinTypeDescriptor> getBuiltinTypeInfo(const TypeRef *TR);
+
+    void populateFieldTypeInfoCacheWithReflectionAtIndex(size_t Index);
+
+    llvm::Optional<RemoteRef<FieldDescriptor>>
+    findFieldDescriptorAtIndex(size_t Index, const std::string &MangledName);
+
+    llvm::Optional<RemoteRef<FieldDescriptor>>
+    getFieldDescriptorFromExternalCache(const std::string &MangledName);
+
+    bool reflectionNameMatches(RemoteRef<char> reflectionName,
+                               StringRef searchName);
+
+    std::vector<ReflectionInfo> ReflectionInfos;
+
+    /// Indexes of Reflection Infos we've already processed.
+    llvm::DenseSet<size_t> ProcessedReflectionInfoIndexes;
+
+    /// Cache for field info lookups.
+    std::unordered_map<std::string, RemoteRef<FieldDescriptor>>
+        FieldTypeInfoCache;
+
+    /// Cache for normalized reflection name lookups.
+    std::unordered_map<uint64_t /* remote address */,
+                       llvm::Optional<std::string>>
+        NormalizedReflectionNameCache;
+
+    /// Cache for built-in type descriptor lookups.
+    std::unordered_map<std::string /* normalized name */,
+                       RemoteRef<BuiltinTypeDescriptor>>
+        BuiltInTypeDescriptorCache;
+    ///
+    /// Cache for associated type lookups.
+    std::unordered_map<TypeRefID, const TypeRef *, TypeRefID::Hash,
+                       TypeRefID::Equal>
+        AssociatedTypeCache;
+
+    /// The index of the last ReflectionInfo cached by
+    /// BuiltInTypeDescriptorCache.
+    uint32_t NormalizedReflectionNameCacheLastReflectionInfoCache = 0;
+
+    MetadataSourceBuilder MSB;
+
+    TypeRefBuilder &Builder;
+
+    remote::ExternalTypeRefCache *ExternalTypeRefCache = nullptr;
+
+  public:
+    ///
+    /// Dumping typerefs, field declarations, builtin types, captures,
+    /// multi-payload enums
+    ///
+    void dumpTypeRef(RemoteRef<char> MangledName, std::ostream &stream,
+                     bool printTypeName = false);
+    FieldTypeCollectionResult
+    collectFieldTypes(llvm::Optional<std::string> forMangledTypeName);
+    void dumpFieldSection(std::ostream &stream);
+    void dumpBuiltinTypeSection(std::ostream &stream);
+    void dumpCaptureSection(std::ostream &stream);
+    void dumpMultiPayloadEnumSection(std::ostream &stream);
+
+    template <template <typename Runtime> class ObjCInteropKind,
+              unsigned PointerSize>
+    AssociatedTypeCollectionResult
+    collectAssociatedTypes(llvm::Optional<std::string> forMangledTypeName) {
```

Comment:
does this have to be in the .h file, or could this be moved into a .cpp file?

Reply:
(and maybe the answer is yes, because of the template args)

Reply:
Yes, it needs to be in the .h due to the template args